## Gather

In [1]:
import pandas as pd

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess

In [3]:
patients

patient_id assigned_sex  given_name       surname  \
0             1       female         Zoe       Wellish   
1             2       female      Pamela          Hill   
2             3         male         Jae        Debord   
3             4         male        Liêm          Phan   
4             5         male         Tim       Neudorf   
5             6         male      Rafael         Costa   
6             7       female        Mary         Adams   
7             8       female      Xiuxiu         Chang   
8             9         male       Dsvid    Gustafsson   
9            10       female      Sophie       Cabrera   
10           11       female       Sandy    Gunnarsson   
11           12         male   Abdul-Nur           Isa   
12           13         male  Omeokachie      Ibeamaka   
13           14       female    Anenechi         Chidi   
14           15       female        Asia       Woźniak   
15           16         male       Søren          Lund   
16           17       female         Tám          Liễu   
17           18       female     Roxanne     Andreyeva   
18           19         male     William         Oates   
19           20         male         Zak         Kelly   
20           21       female       Sofia       Karlsen   
21           22         male      Samúel  Guðbrandsson   
22           23         male      Manchu            Su   
23           24         male       Lovre         Galić   
24           25         male       Jakob      Jakobsen   
25           26         male      Gregor          Bole   
26           27       female        Ella          Lund   
27           28         male      Joseph        Tucker   
28           29         male      Robert          Wolf   
29           30         male        Jake      Jakobsen   
..          ...          ...         ...           ...   
473         474       female        Kate     Wilkinson   
474         475       female   Esperanza      Labrosse   
475         476         male       Malik       Vaneker   
476         477       female       Berta    Napolitani   
477         478         male     Juliusz      Majewski   
478         479       female      Edelma   Villalpando   
479         480         male        Tapa   Arsanukayev   
480         481         male      Nasser       Mansour   
481         482         male     Michael    Kristensen   
482         483         male       Diogo         Souza   
483         484       female       Angel         Grant   
484         485         male     Placido       Udinesi   
485         486         male      Trifon      Izmailov   
486         487         male      Samuel          Blix   
487         488         male        Ivar       Löfgren   
488         489         male        Mika    Martinsson   
489         490       female     Jasmine         Sykes   
490         491         male     Jackson       Addison   
491         492       female     Vanessa      Ferguson   
492         493         male       Poldi           Tar   
493         494       female         Fen          Chin   
494         495       female      Sirkka     Piirainen   
495         496         male      Hajime       Tsukada   
496         497         male   Alexander        Hueber   
497         498         male    Masataka      Murakami   
498         499         male     Mustafa     Lindström   
499         500         male       Ruman       Bisliev   
500         501       female       Jinke     de Keizer   
501         502       female     Chidalu   Onyekaozulu   
502         503         male         Pat       Gersten   

                        address               city       state  zip_code  \
0          576 Brown Bear Drive  Rancho California  California   92390.0   
1     2370 University Hill Road          Armstrong    Illinois   61812.0   
2         1493 Poling Farm Road               York    Nebraska   68467.0   
3           2335 Webster Street         Woodbridge          NJ    7095.0   
4          1428 Turkey Pen L

In [4]:
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0       veronika      jindrová  41u - 48u          -         7.63       7.20   
1         elliot    richardson          -  40u - 45u         7.56       7.09   
2       yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3           skye    gormanston  33u - 36u          -         7.97       7.62   
4         alissa        montez          -  33u - 29u         7.78       7.46   
5        jasmine         sykes          -  42u - 44u         7.56       7.18   
6         sophia        haugen  37u - 42u          -         7.65       7.27   
7          eddie        archer  31u - 38u          -         7.89       7.55   
8          saber        ménard          -  54u - 54u         8.08       7.70   
9           asia       woźniak  30u - 36u          -         7.76       7.37   
10        joseph           day  29u - 36u          -         7.70       7.19   
11     kristiina        hyypiä          -  36u - 38u         7.87       7.49   
12       roxanne     andreyeva  29u - 38u          -         9.54       9.14   
13        gregor          bole          -  47u - 45u         7.61       7.16   
14        simone  baumgaertner  27u - 37u          -         7.74       7.30   
15          enco        žibrik  55u - 68u          -         7.78       7.34   
16       camilla      zaitseva  28u - 37u          -         7.53       7.13   
17          gina          cain          -  36u - 36u         7.88       7.40   
18    addolorata      lombardi          -  49u - 46u         7.75       7.33   
19        khalid      johnsrud          -  54u - 54u         8.35       7.94   
20          mile        stanić          -  47u - 48u         7.66       7.24   
21         tekla       walczak  29u - 39u          -         7.61       7.29   
22   brancaleone         russo  53u - 60u          -         8.61       8.18   
23      chiemela     tobeolisa          -  43u - 47u         7.59       7.17   
24          isac          berg  31u - 41u          -         9.68       9.29   
25        benoît        bonami          -  44u - 43u         9.82       9.40   
26        suhaim         rahal          -  49u - 47u         7.94       7.50   
27        mizuki         iwata          -  45u - 46u         7.70       7.23   
28       clinton        miller  42u - 51u          -         7.79       7.40   
29        eugene       mironov  42u - 49u          -         7.81       7.48   
..           ...           ...        ...        ...          ...        ...   
250         chen           yao          -  56u - 57u         7.90       7.51   
251        aksel   vestergaard          -  42u - 38u         9.62       9.29   
252        ellen         luman          -  40u - 39u         9.27       8.77   
253       albino     schiavone  35u - 43u          -         7.56       7.15   
254         jose         combs          -  39u - 36u         7.89       7.42   
255       jia li          teng  48u - 54u          -         7.66       7.32   
256        ilija        horvat  42u - 50u          -         7.77       7.38   
257     mathilde      nørgaard          -  27u - 28u         8.50       8.10   
258       csilla      herczegh          -  43u - 46u         7.71       7.27   
259      aaliyah          rice          -  31u - 31u         7.64       7.33   
260        david      beauvais          -  26u - 23u         7.87       7.47   
261     caroline        shuler          -  50u - 54u         7.63       7.27   
262         alex      crawford  51u - 62u          -         7.69       7.30   
263      rebecca      jephcott  53u - 63u          -         7.96       7.57   
264   chukwumoge     ogochukwu          -  41u - 39u         7.95       7.56   
265       fearne      mcgregor          -  27u - 29u         7.83       7.48   
266       ursula         freud  42u - 54u          -         7.75       7.46   
267         leon        scholz          -  38u - 32u         7.72       7.29   
268   

In [5]:
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

In [6]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [7]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [8]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [9]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [10]:
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

In [11]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [12]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [13]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [14]:
patients.sample(5)

patient_id assigned_sex given_name        surname  \
20           21       female      Sofia        Karlsen   
73           74       female      Žarka            Rap   
261         262         male     Jeremy        Montagu   
290         291       female        Una  Traustadóttir   
377         378       female    Cumelén     Miramontes   

                    address     city     state  zip_code        country  \
20       2931 Romano Street  Whitman        MA    2382.0  United States   
73      2661 Russell Street  Foxboro        MA    2035.0  United States   
261        4988 Lynn Street   Boston        MA    2110.0  United States   
290       1027 Tenmile Road  Bedford        MA    1730.0  United States   
377  4213 Isaacs Creek Road   Tolono  Illinois   61880.0  United States   

                                           contact   birthdate  weight  \
20         SofiaTKarlsen@teleworm.us1 781 447 1763   9/24/1934   153.1   
73              978-969-6753ZarkaRap@superrito.com  12/28/1995   139.3   
261          JeremyMontagu@teleworm.us617-297-0387   6/23/1982   192.7   
290       781-739-0244UnaTraustadottir@armyspy.com    2/1/1930   152.3   
377  217-485-5673CumelenMiramontesAcosta@gustr.com   1/30/1936   130.0   

     height   bmi  
20       66  24.7  
73       68  21.2  
261      74  24.7  
290      59  30.8  
377      66  21.0

In [15]:
patients.surname.value_counts()

Doe            6
Taylor         3
Jakobsen       3
Souza          2
Aranda         2
Cindrić        2
Collins        2
Batukayev      2
Silva          2
Kowalczyk      2
Bùi            2
Berg           2
Tucker         2
Lương          2
Correia        2
Lund           2
Woźniak        2
Lâm            2
Ogochukwu      2
Tạ             2
Liễu           2
Parker         2
Schiavone      2
Kadyrov        2
Hueber         2
Johnson        2
Grímsdóttir    2
Cabrera        2
Dratchev       2
Gersten        2
              ..
van de Wiel    1
Nyborg         1
Eidem          1
Filípek        1
Poulsen        1
Tikhonov       1
Quintanilla    1
Koffeman       1
Rice           1
Luman          1
Vestergaard    1
Hsueh          1
Sakai          1
Hill           1
Kos            1
Mathieu        1
Christie       1
Bjarkason      1
Ingason        1
Onyekaozulu    1
Amari          1
Glaser         1
Német          1
Khoury         1
Chandler       1
Nakahara       1
Zsinkó         1
Achebe        

In [16]:
patients.address.value_counts()

123 Main Street             6
2476 Fulton Street          2
2778 North Avenue           2
648 Old Dear Lane           2
1854 Kyle Street            1
2886 Straford Park          1
4220 Simpson Square         1
3683 Gorby Lane             1
2641 Michael Street         1
3216 Lodgeville Road        1
1886 Bicetown Road          1
1895 Hart Ridge Road        1
230 Lady Bug Drive          1
846 Copperhead Road         1
4320 Rardin Drive           1
4040 Linda Street           1
602 Tator Patch Road        1
3558 Longview Avenue        1
1370 Flint Street           1
3857 Straford Park          1
2813 Frederick Street       1
631 Isaacs Creek Road       1
2324 Benson Street          1
4033 White Avenue           1
142 Broad Street            1
3084 Blue Spruce Lane       1
826 Broad Street            1
108 Griffin Street          1
4789 Devils Hill Road       1
2913 Rogers Street          1
                           ..
2338 Virginia Street        1
1386 Farland Street         1
541 Green 

In [17]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [18]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
       ...  
332    224.0
252    224.2
12     224.2
222    224.8
166    225.3
111    225.9
101    226.2
150    226.6
352    227.7
428    227.7
88     227.7
13     228.4
339    229.0
182    230.3
121    230.8
257    231.7
395    231.9
246    232.1
219    237.8
11     238.7
50     238.9
441    239.1
499    239.6
439    242.0
487    242.4
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [19]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [20]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [21]:
sum(treatments.auralin.isnull())

0

In [22]:
sum(treatments.novodra.isnull())

0

#### Quality
##### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

##### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `adverse_reactions` table
- Lowercase given names and surnames

#### Tidiness
- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- Adverse reaction should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

## Clean

In [23]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### Missing Data

<font color='red'>Complete the following two "Missing Data" **Define, Code, and Test** sequences after watching the *"Address Missing Data First"* video.</font>

#### `treatments`: Missing records (280 instead of 350)

##### Define
*Note: the missing `treatments` records are stored in a file named `treatments_cut.csv`, which you can see in this Jupyter Notebook's dashboard (click the **jupyter** logo in the top lefthand corner of this Notebook). Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html) for the function used in the solution.*

Append the missing **treatments** records stored in `treatments_cut.csv` to `treatments_clean` DataFrame

##### Code

In [24]:
# Read missing treatments into a DataFrame
treatments_cut = pd.read_csv('treatments_cut.csv')

# Append missing treatments records into treatments_clean DataFrame
treatments_clean = treatments_clean.append(treatments_cut, ignore_index=True)

##### Test

In [25]:
# Your testing code here
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 7 columns):
given_name      350 non-null object
surname         350 non-null object
auralin         350 non-null object
novodra         350 non-null object
hba1c_start     350 non-null float64
hba1c_end       350 non-null float64
hba1c_change    213 non-null float64
dtypes: float64(3), object(4)
memory usage: 19.2+ KB


#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)
*Note: the "Inaccurate HbA1c changes (leading 4s mistaken as 9s)" observation, which is an accuracy issue and not a completeness issue, is included in this header because it is also fixed by the cleaning operation that fixes the missing "Missing HbA1c changes" observation. Multiple observations in one **Define, Code, and Test** header occurs multiple times in this notebook.*

##### Define
Fix missing and inaccurate HbA1c changes by subtracting `hba1c_start` value from `hba1c_end` value inplace.  

##### Code

In [26]:
# Fix missing and inaccurate HbA1c changes column
treatments_clean['hba1c_change'] = (treatments_clean['hba1c_end'] - 
                                    treatments_clean['hba1c_start'])

##### Test

In [27]:
# Testing for missing HbA1c changes
treatments_clean['hba1c_change'].isnull().sum()

0

In [28]:
# Testing for inaccurate HbA1c change values
assert ((treatments_clean['hba1c_end'] - treatments_clean['hba1c_start'])
        .equals(treatments_clean['hba1c_change']))

In [29]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 7 columns):
given_name      350 non-null object
surname         350 non-null object
auralin         350 non-null object
novodra         350 non-null object
hba1c_start     350 non-null float64
hba1c_end       350 non-null float64
hba1c_change    350 non-null float64
dtypes: float64(3), object(4)
memory usage: 19.2+ KB


### Tidiness

<font color='red'>Complete the following four "Tidiness" **Define, Code, and Test** sequences after watching the *"Cleaning for Tidiness"* video.</font>

#### Contact column in `patients` table contains two variables: phone number and email

##### Define
*Your definition here. Hint 1: use regular expressions with pandas' [`str.extract` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.extract.html). Here is an amazing [regex tutorial](https://regexone.com/). Hint 2: [various phone number regex patterns](https://stackoverflow.com/questions/16699007/regular-expression-to-match-standard-10-digit-phone-number). Hint 3: [email address regex pattern](http://emailregex.com/), which you might need to modify to distinguish the email from the phone number.*

##### Code

In [30]:
# Your cleaning code here
re_phone = r"^(\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4}$"
re_email = r"(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$)"
s = patients['contact'].str.extract(r"(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$)", expand=True)

##### Test

In [31]:
# Your testing code here
s

0
0           951-719-9170ZoeWellish@superrito.com
1                                            NaN
2               402-363-6804JaeMDebord@gustr.com
3                                            NaN
4                334-515-7487TimNeudorf@cuvox.de
5       386-334-5237RafaelCardosoCosta@gustr.com
6              775-533-5933MaryBAdams@einrot.com
7                                            NaN
8        816-265-9578DavidGustafsson@armyspy.com
9                                            NaN
10        906-478-8949SandyGunnarsson@dayrep.com
11                                           NaN
12     OmeokachieIbeamaka@einrot.com434-509-2614
13                                           NaN
14             AsiaWozniak@rhyta.com918-712-3469
15               276-225-1955SrenFLund@gustr.com
16                                           NaN
17      RoxanneAndreyeva@armyspy.com678-829-8578
18         406-775-2696WilliamVOates@armyspy.com
19                                           NaN
20                                           NaN
21     973-445-5341SamuelGubrandsson@teleworm.us
22               914-745-6108ManchuSu@einrot.com
23                                           NaN
24                                           NaN
25              GregorBole@gustr.com505-828-4955
26              309-671-8852EllaLund@armyspy.com
27           985-814-7603JosephNTucker@rhyta.com
28                                           NaN
29                                           NaN
..                                           ...
473                                          NaN
474    EsperanzaLabrosse@armyspy.com678-263-3564
475       MalikVaneker@superrito.com650-962-7179
476        267-972-3749BertaNapolitani@rhyta.com
477                                          NaN
478                                          NaN
479                                          NaN
480                                          NaN
481                                          NaN
482  361-693-4960DiogoBarrosSouza@jourrapide.com
483           731-577-0292AngelGrant@fleckens.hu
484        336-697-2005PlacidoUdinesi@dayrep.com
485                                          NaN
486            254-681-4504SamuelBlix@dayrep.com
487                                          NaN
488       352-453-4601MikaMartinsson@armyspy.com
489      JasmineSykes@jourrapide.com925-283-5425
490       914-636-9304JacksonAddison@armyspy.com
491   210-222-8684VanessaFerguson@jourrapide.com
492           714-496-2264TarPoldi@superrito.com
493                                          NaN
494                                          NaN
495         650-570-4896HajimeTsukada@dayrep.com
496                                          NaN
497                                          NaN
498  207-477-0579MustafaLindstrom@jourrapide.com
499           928-284-4492RumanBisliev@gustr.com
500        816-223-6007JinkedeKeizer@teleworm.us
501                                          NaN
502         PatrickGersten@rhyta.com402-848-4923

[503 rows x 1 columns]

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

##### Define
Transform columns auralin and novodra into values under treatments column. Create two new columns (variables) to store start and end doses. Remove string 'u' from data and convert to integers.

Hint: use pandas' [melt function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html) and [`str.split()` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.split.html). Here is an excellent [`melt` tutorial](https://deparkes.co.uk/2016/10/28/reshape-pandas-data-with-melt/).*

##### Code

In [32]:
# Your cleaning code here
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'], var_name='treatment', value_vars=['auralin', 'novodra'])

# Drop rows which treatment values are '-'
index_names = treatments_clean[treatments_clean['value'] == '-'].index
treatments_clean.drop(index_names, inplace=True)

treatments_clean.reset_index(drop=True,inplace=True)

# Split value column by '-' delimiter
treatments_clean[['start_dose', 'end_dose']] = treatments_clean.value.str.split('-', expand=True)

# Remove 'value' column
treatments_clean.drop(columns=['value'], inplace=True)

# Remove character 'u' from start_dose and end_dose columns and convert them to integer
treatments_clean['start_dose'] = treatments_clean['start_dose'].str.replace('u', '').astype(int)
treatments_clean['end_dose'] = treatments_clean['end_dose'].str.replace('u', '').astype(int)

# Reorder columns to original structure
ordered_columns = (['given_name', 'surname', 'treatment', 'start_dose', 'end_dose', 
                    'hba1c_start', 'hba1c_end', 'hba1c_change'])

treatments_clean = treatments_clean[ordered_columns]

##### Test

In [33]:
# Your testing code here
treatments_clean

given_name        surname treatment  start_dose  end_dose  hba1c_start  \
0        veronika       jindrová   auralin          41        48         7.63   
1            skye     gormanston   auralin          33        36         7.97   
2          sophia         haugen   auralin          37        42         7.65   
3           eddie         archer   auralin          31        38         7.89   
4            asia        woźniak   auralin          30        36         7.76   
5          joseph            day   auralin          29        36         7.70   
6         roxanne      andreyeva   auralin          29        38         9.54   
7          simone   baumgaertner   auralin          27        37         7.74   
8            enco         žibrik   auralin          55        68         7.78   
9         camilla       zaitseva   auralin          28        37         7.53   
10          tekla        walczak   auralin          29        39         7.61   
11    brancaleone          russo   auralin          53        60         8.61   
12           isac           berg   auralin          31        41         9.68   
13        clinton         miller   auralin          42        51         7.79   
14         eugene        mironov   auralin          42        49         7.81   
15    szilveszter          totth   auralin          35        39         7.70   
16      alexander      mathiesen   auralin          47        58         7.96   
17            chỉ            lâm   auralin          45        48         7.68   
18      władysław      wieczorek   auralin          24        37         7.92   
19       kristján        ingason   auralin          44        55         7.92   
20         marija       grubišić   auralin          37        43         7.53   
21          sauli     koivuniemi   auralin          43        47         7.67   
22        mariana          souza   auralin          36        42         7.86   
23     kristoffer      martinsen   auralin          29        37         9.18   
24             mỹ          quynh   auralin          57        64         7.61   
25           oles        zhdanov   auralin          54        67         7.52   
26        triana.       terrazas   auralin          34        42         7.71   
27         gabryŝ    tomaszewski   auralin          29        37         7.87   
28      leixandre         alanis   auralin          61        67         7.74   
29   onyekachukwu         obinna   auralin          37        46         7.58   
..            ...            ...       ...         ...       ...          ...   
320          jane        citizen   novodra          37        38         7.98   
321        angela     lavrentyev   novodra          28        24         7.61   
322        edelma    villalpando   novodra          24        26         7.99   
323        annika          vaara   novodra          20        21         7.73   
324         chiho           higa   novodra          46        46         7.71   
325     beatrycze        woźniak   novodra          26        27         7.54   
326      miłosław     wiśniewski   novodra          34        33         7.51   
327       firenze          fodor   novodra          30        35         7.89   
328           zoe        wellish   novodra          33        33         7.71   
329           una  traustadóttir   novodra          35        34         8.00   
330         luboš          pecha   novodra          30        27         7.79   
331         meaza         brhane   novodra          37        41         7.70   
332         adlan       shishani   novodra          43        40         7.84   
333         sofia      hermansen   novodra          34        34         8.90   
334         guðni      heimisson   novodra          40        36         7.64   
335       eufemio        rosario   novodra          37        40         7.54   
336      dalmacia         madrid   novodra          26        23         7.67   
337         daimy          tromp   nov

#### Adverse reaction should be part of the `treatments` table

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for the function used in the solution.*

##### Code

In [34]:
# Your cleaning code here
treatments_clean = pd.merge(treatments_clean, adverse_reactions_clean, on=('given_name', 'surname'), how='left')

##### Test

In [35]:
# Your testing code here
treatments_clean

given_name        surname treatment  start_dose  end_dose  hba1c_start  \
0        veronika       jindrová   auralin          41        48         7.63   
1            skye     gormanston   auralin          33        36         7.97   
2          sophia         haugen   auralin          37        42         7.65   
3           eddie         archer   auralin          31        38         7.89   
4            asia        woźniak   auralin          30        36         7.76   
5          joseph            day   auralin          29        36         7.70   
6         roxanne      andreyeva   auralin          29        38         9.54   
7          simone   baumgaertner   auralin          27        37         7.74   
8            enco         žibrik   auralin          55        68         7.78   
9         camilla       zaitseva   auralin          28        37         7.53   
10          tekla        walczak   auralin          29        39         7.61   
11    brancaleone          russo   auralin          53        60         8.61   
12           isac           berg   auralin          31        41         9.68   
13        clinton         miller   auralin          42        51         7.79   
14         eugene        mironov   auralin          42        49         7.81   
15    szilveszter          totth   auralin          35        39         7.70   
16      alexander      mathiesen   auralin          47        58         7.96   
17            chỉ            lâm   auralin          45        48         7.68   
18      władysław      wieczorek   auralin          24        37         7.92   
19       kristján        ingason   auralin          44        55         7.92   
20         marija       grubišić   auralin          37        43         7.53   
21          sauli     koivuniemi   auralin          43        47         7.67   
22        mariana          souza   auralin          36        42         7.86   
23     kristoffer      martinsen   auralin          29        37         9.18   
24             mỹ          quynh   auralin          57        64         7.61   
25           oles        zhdanov   auralin          54        67         7.52   
26        triana.       terrazas   auralin          34        42         7.71   
27         gabryŝ    tomaszewski   auralin          29        37         7.87   
28      leixandre         alanis   auralin          61        67         7.74   
29   onyekachukwu         obinna   auralin          37        46         7.58   
..            ...            ...       ...         ...       ...          ...   
320          jane        citizen   novodra          37        38         7.98   
321        angela     lavrentyev   novodra          28        24         7.61   
322        edelma    villalpando   novodra          24        26         7.99   
323        annika          vaara   novodra          20        21         7.73   
324         chiho           higa   novodra          46        46         7.71   
325     beatrycze        woźniak   novodra          26        27         7.54   
326      miłosław     wiśniewski   novodra          34        33         7.51   
327       firenze          fodor   novodra          30        35         7.89   
328           zoe        wellish   novodra          33        33         7.71   
329           una  traustadóttir   novodra          35        34         8.00   
330         luboš          pecha   novodra          30        27         7.79   
331         meaza         brhane   novodra          37        41         7.70   
332         adlan       shishani   novodra          43        40         7.84   
333         sofia      hermansen   novodra          34        34         8.90   
334         guðni      heimisson   novodra          40        36         7.64   
335       eufemio        rosario   novodra          37        40         7.54   
336      dalmacia         madrid   novodra          26        23         7.67   
337         daimy          tromp   nov

In [36]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 349
Data columns (total 9 columns):
given_name          350 non-null object
surname             350 non-null object
treatment           350 non-null object
start_dose          350 non-null int64
end_dose            350 non-null int64
hba1c_start         350 non-null float64
hba1c_end           350 non-null float64
hba1c_change        350 non-null float64
adverse_reaction    35 non-null object
dtypes: float64(3), int64(2), object(4)
memory usage: 27.3+ KB


#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

In [37]:
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for one function used in the solution and [tutorial](http://erikrood.com/Python_References/dropping_rows_cols_pandas.html) for another function used in the solution.*

##### Code

In [38]:
# Capitalize first character on given_name and surname columns
treatments_clean['given_name'] = treatments_clean['given_name'].str.title()
treatments_clean['surname'] = treatments_clean['surname'].str.title() 

# Merge patient_id column from patients_clean into treatments_clean
treatments_clean = pd.merge(treatments_clean, patients_clean[['patient_id', 'given_name', 'surname']], on=('given_name', 'surname'), how='left')

# Remove given_name and surname columns
treatments_clean.drop(['given_name', 'surname'], axis=1, inplace=True)

# Reorder columns in treatments_clean DataFrame
ordered_columns = (['patient_id', 'treatment', 'start_dose', 'end_dose', 'hba1c_start',
                    'hba1c_end', 'hba1c_change', 'adverse_reaction'])

treatments_clean = treatments_clean[ordered_columns]

In [39]:
# Capitalize first character on given_name and surname columns
adverse_reactions_clean['given_name'] = adverse_reactions_clean['given_name'].str.title()
adverse_reactions_clean['surname'] = adverse_reactions_clean['surname'].str.title()

# Merge patient_id column from patients_clean into adverse_reactions_clean
adverse_reactions_clean = pd.merge(adverse_reactions_clean, 
                                   patients_clean[['patient_id', 'given_name', 'surname']], 
                                   on=('given_name', 'surname'), how='left')

# Remove given_name and surname_columnss
adverse_reactions_clean.drop(['given_name', 'surname'], axis=1, inplace=True)

# Reorder columns in adverse_reactions_clean DataFrame
adverse_reactions_clean = adverse_reactions_clean[['patient_id', 'adverse_reaction']]

##### Test

In [40]:
# Your testing code here
treatments_clean[treatments_clean['patient_id'].isnull()]

patient_id treatment  start_dose  end_dose  hba1c_start  hba1c_end  \
113         NaN   auralin          33        39         7.55       7.26   
214         NaN   novodra          44        43         9.87       9.48   
244         NaN   novodra          33        34         7.72       7.28   
247         NaN   novodra          31        29         7.94       7.51   
277         NaN   novodra          40        40         7.57       7.20   
311         NaN   novodra          27        29         7.83       7.48   

     hba1c_change adverse_reaction  
113         -0.29              NaN  
214         -0.39              NaN  
244         -0.44              NaN  
247         -0.43              NaN  
277         -0.37              NaN  
311         -0.35              NaN

In [41]:
adverse_reactions_clean

patient_id           adverse_reaction
0          477  injection site discomfort
1          128               hypoglycemia
2           70               hypoglycemia
3          270                      cough
4          403          throat irritation
5          490               hypoglycemia
6           55               hypoglycemia
7           38               hypoglycemia
8           52               hypoglycemia
9          376  injection site discomfort
10         408                   headache
11         259                      cough
12          12               hypoglycemia
13         332  injection site discomfort
14         209               hypoglycemia
15         452                     nausea
16          25               hypoglycemia
17         153                     nausea
18         354               hypoglycemia
19         143                   headache
20          14               hypoglycemia
21         373  injection site discomfort
22         336  injection site discomfort
23         360               hypoglycemia
24         255  injection site discomfort
25          94               hypoglycemia
26         451          throat irritation
27         460               hypoglycemia
28           8               hypoglycemia
29         277               hypoglycemia
30         178               hypoglycemia
31         394                   headache
32         249               hypoglycemia
33         254               hypoglycemia

### Quality

<font color='red'>Complete the remaining "Quality" **Define, Code, and Test** sequences after watching the *"Cleaning for Quality"* video.</font>

#### Zip code is a float not a string and Zip code has four digits sometimes

##### Define
Convert zip code column into string and pad zero at start of string. 

Hint: see the "Data Cleaning Process" page.*

##### Code

In [46]:
# Convert float to string and remove .0 from string
patients_clean['zip_code'] = patients_clean['zip_code'].astype(str).str.replace(r'\.0', '')

# Pad leading zero to zip code colum
patients_clean['zip_code'] = [str(i).rjust(5, '0') for i in patients_clean['zip_code']]

##### Test

In [47]:
# Your testing code here
patients_clean

patient_id assigned_sex  given_name       surname  \
0             1       female         Zoe       Wellish   
1             2       female      Pamela          Hill   
2             3         male         Jae        Debord   
3             4         male        Liêm          Phan   
4             5         male         Tim       Neudorf   
5             6         male      Rafael         Costa   
6             7       female        Mary         Adams   
7             8       female      Xiuxiu         Chang   
8             9         male       Dsvid    Gustafsson   
9            10       female      Sophie       Cabrera   
10           11       female       Sandy    Gunnarsson   
11           12         male   Abdul-Nur           Isa   
12           13         male  Omeokachie      Ibeamaka   
13           14       female    Anenechi         Chidi   
14           15       female        Asia       Woźniak   
15           16         male       Søren          Lund   
16           17       female         Tám          Liễu   
17           18       female     Roxanne     Andreyeva   
18           19         male     William         Oates   
19           20         male         Zak         Kelly   
20           21       female       Sofia       Karlsen   
21           22         male      Samúel  Guðbrandsson   
22           23         male      Manchu            Su   
23           24         male       Lovre         Galić   
24           25         male       Jakob      Jakobsen   
25           26         male      Gregor          Bole   
26           27       female        Ella          Lund   
27           28         male      Joseph        Tucker   
28           29         male      Robert          Wolf   
29           30         male        Jake      Jakobsen   
..          ...          ...         ...           ...   
473         474       female        Kate     Wilkinson   
474         475       female   Esperanza      Labrosse   
475         476         male       Malik       Vaneker   
476         477       female       Berta    Napolitani   
477         478         male     Juliusz      Majewski   
478         479       female      Edelma   Villalpando   
479         480         male        Tapa   Arsanukayev   
480         481         male      Nasser       Mansour   
481         482         male     Michael    Kristensen   
482         483         male       Diogo         Souza   
483         484       female       Angel         Grant   
484         485         male     Placido       Udinesi   
485         486         male      Trifon      Izmailov   
486         487         male      Samuel          Blix   
487         488         male        Ivar       Löfgren   
488         489         male        Mika    Martinsson   
489         490       female     Jasmine         Sykes   
490         491         male     Jackson       Addison   
491         492       female     Vanessa      Ferguson   
492         493         male       Poldi           Tar   
493         494       female         Fen          Chin   
494         495       female      Sirkka     Piirainen   
495         496         male      Hajime       Tsukada   
496         497         male   Alexander        Hueber   
497         498         male    Masataka      Murakami   
498         499         male     Mustafa     Lindström   
499         500         male       Ruman       Bisliev   
500         501       female       Jinke     de Keizer   
501         502       female     Chidalu   Onyekaozulu   
502         503         male         Pat       Gersten   

                        address               city       state zip_code  \
0          576 Brown Bear Drive  Rancho California  California    92390   
1     2370 University Hill Road          Armstrong    Illinois    61812   
2         1493 Poling Farm Road               York    Nebraska    68467   
3           2335 Webster Street         Woodbridge          NJ    07095   
4          1428 Turkey Pen Lane  

#### Tim Neudorf height is 27 in instead of 72 in

##### Define
Query DataFrame for surname equals Neudorf and replace height.

##### Code

In [58]:
# Your cleaning code here
patients_clean.loc[patients_clean['surname'] == 'Neudorf', ['height']] = 72

##### Test

In [59]:
# Your testing code here
patients_clean.loc[patients_clean['surname'] == 'Neudorf']

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state zip_code        country                          contact  birthdate  \
4    AL    36303  United States  334-515-7487TimNeudorf@cuvox.de  2/18/1928   

   weight  height   bmi  
4   192.3      72  26.1

#### Full state names sometimes, abbreviations other times

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_apply_operations_to_dataframes.html) for method used in solution.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Dsvid Gustafsson

##### Define
*Your definition here.*

##### Code

In [61]:
# Your cleaning code here
patients_clean.loc[patients_clean['given_name'] == 'Dsvid', ['given_name']] = 'David'

##### Test

In [63]:
# Your testing code here
patients_clean.loc[patients_clean['surname'] == 'Gustafsson']

patient_id assigned_sex given_name     surname             address  \
8           9         male      David  Gustafsson  1790 Nutter Street   

          city state zip_code        country  \
8  Kansas City    MO    64105  United States   

                                   contact birthdate  weight  height   bmi  
8  816-265-9578DavidGustafsson@armyspy.com  3/6/1937   163.9      66  26.5

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

##### Define
*Your definition here. Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.astype.html) for one method used in solution, [documentation page](http://pandas.pydata.org/pandas-docs/version/0.20/generated/pandas.to_datetime.html) for one function used in the solution, and [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.strip.html) for another method used in the solution.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Multiple phone number formats

##### Define
*Your definition here. Hint: helpful [Stack Overflow answer](https://stackoverflow.com/a/123681).*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Default John Doe data

##### Define
*Your definition here. Recall that it is assumed that the data that this John Doe data displaced is not recoverable.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Multiple records for Jakobsen, Gersten, Taylor

##### Define
*Your definition here.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### kgs instead of lbs for Zaitseva weight

##### Define
*Your definition here.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here